USHIFT-1467: Upgradeability contributor's doc (#2099)
openshift/microshift: docs/contributor/updateability.md

# Upgrading MicroShift

Following document is meant to be a contributor's introduction
to the feature of MicroShift's updateability.

## On-disk artifacts

New on-disk artifacts were added:
- `/var/lib/microshift-backups/`
    - `health.json`
    - backup directories
- `/var/lib/microshift/version`

### `/var/lib/microshift-backups/`

New, additional directory for keeping health and backup related artifacts.
For ostree systems it's a non-configurable directory where backups are created.

> For non-ostree systems it's a default (configurable) directory for backups.

Reviewer: I didn't realize this was configurable. Why is it configurable?

Author: It made sense since it's a cli command. If this is manual process, we don't really care where user will store it.

```
Usage:
  microshift admin data [command]

Available Commands:
  backup      Backup MicroShift data

Flags:
  -h, --help             help for data
      --name string      Backup name (default "4.15__20230721_072807")
      --storage string   Directory with backups (default "/var/lib/microshift-backups")
```

I just noticed that we missed command to restore :smile:

Author: Do you think we should change the cli? From `microshift admin data {backup,restore}` to something more concise?

Author: Changing in https://github.com/openshift/microshift/pull/2108

> #### `health.json` [ostree-only]

Reviewer: Greenboot runs on systems without ostree. Automated rollback is only supported with ostree, but some of the other greenboot features, especially reporting system health holistically, are supported. So, this file might exist on a regular system, too.

Author: We don't create it on non-ostree ([ref](https://github.com/openshift/microshift/blob/72e8e0aab81342e1db9f773dc5f1d67f0407540c/packaging/greenboot/microshift_set_healthy.sh#L13-L16)):
```sh
if [ ! -f /run/ostree-booted ]; then
    echo "System is not booted with ostree"
    exit 0
fi
```

  - Expected values: `healthy` or `unhealthy`
- `deployment_id`: OSTree deployment ID.
  - Obtained with a command: `rpm-ostree status --json | jq '.deployments[].id'`
- `boot_id`: Boot ID generated by kernel.
  - Obtained with command: `tr -d '-' < /proc/sys/kernel/random/boot_id`.
  - Hyphens are removed to match format used by `journalctl` (see `journalctl --list-boots --reverse`)

#### Backup directories

> Note: Only ostree-based systems perform automated backups and restores.

`/var/lib/microshift-backups/` also stores backups of MicroShift data.
Backups are created by copying data directory with `--reflink`
option to leverage Copy-on-Write functionality.

Naming schema of backup directories: `deployment-id_boot-id[_unhealthy]`.

Suffix `_unhealthy` is used to mark backups of unhealthy systems.
These backups are performed when all existing metadata suggests that it is expected
that MicroShift should wipe the data and start from clean state
(for example "FDO" scenario where unhealthy deployment leaves stale data).

### `/var/lib/microshift/version`

TODO

Schema:
```
MAJOR.MINOR.PATCH
```

All of MAJOR, MINOR, and PATCH are unsigned integers.
For example: `4.14.0`

> ## Greenboot [ostree-only]

Reviewer: Same comment, the greenboot scripts can and might be installed on systems not based on ostree.

Author: Well, yes. But it's usefulness is greatly reduced. Also, you can perform many lifecycle operations on microshift on RPM without rerunning the greenboot.

I'll add a comment

Backup management and version metadata management run
before MicroShift attempts to start all the components and cluster itself.
Storage migration runs when cluster is up and running 
(and in 10 minutes intervals afterwards) because it
requires runtime components to correctly perform its task.

### Automated backup management [ostree-only]

It runs very close to the start of MicroShift's `run` command.

It checks for existence of `/run/ostree-booted` file.
If the file is not present, it is assumed that the system is regular RPM based
and this stage is skipped without an error.

#### Handling unusual scenarios

Then, procedure checks existence of three on-disk artifacts:
- `/var/lib/microshift` (ignores `.nodename`[0]) - referred to as `data` below
- `/var/lib/microshift/version`
- `/var/lib/microshift-backups/health.json`

[0] `.nodename` might be created by the "effective config generation" phase
which happens before the prerun. This can create `.nodename` and result
in `/var/lib/microshift` being present but, beside that one file, empty.

##### `/var/lib/microshift` (data) does not exist

> Lack of `/var/lib/microshift` implies that `/var/lib/microshift/version` does not exist.

> If both `data` and `health.json` are missing, then we assume it's a first run of MicroShift.

Reviewer: nit: Using "`data`" here makes me think that's the name of an actual file, but I think you mean `/var/lib/microshift` == `data`? Maybe use _data_ instead?

# Upgrading MicroShift

Following document is meant to be a contributor's introduction
to the feature of MicroShift's updateability.

## On-disk artifacts

New on-disk artifacts were added:
- `/var/lib/microshift-backups/`
    - `health.json`
    - backup directories
- `/var/lib/microshift/version`

### `/var/lib/microshift-backups/`

New, additional directory for keeping health and backup related artifacts.
For ostree systems it's a non-configurable directory where backups are created.
For non-ostree systems it's a default (configurable) directory for backups.

#### `health.json` [ostree-only]

Contains information about health of the system.
It's updated by green and red script supplied by MicroShift.

Because MicroShift's healthcheck takes some time to assess the state of the MicroShift,

> this file might be updated after couple minutes into system boot.

Reviewer 2: Let's add a command here to display the status of greenboot, so that the user knows when the file is updated.

Schema:
```json
{
    "health": "",
    "deployment_id": "",
    "boot_id": ""
}
```

- `health`: System status according to the greenboot healthchecks.
  - Expected values: `healthy` or `unhealthy`
- `deployment_id`: OSTree deployment ID.
  - Obtained with a command: `rpm-ostree status --json | jq '.deployments[].id'`
- `boot_id`: Boot ID generated by kernel.
  - Obtained with command: `tr -d '-' < /proc/sys/kernel/random/boot_id`.
  - Hyphens are removed to match format used by `journalctl` (see `journalctl --list-boots --reverse`)

#### Backup directories

> Note: Only ostree-based systems perform automated backups and restores.

`/var/lib/microshift-backups/` also stores backups of MicroShift data.
Backups are created by copying data directory with `--reflink`
option to leverage Copy-on-Write functionality.

> Naming schema of backup directories: `deployment-id_boot-id[_unhealthy]`.

Author: Do you find `deploymentID_bootID` less confusing?

> Naming schema of backup directories: `deployment-id_boot-id[_unhealthy]`.

Reviewer: Yes, that works for me.

primarily for ostree-based systems.

For more in-depth information about greenboot and how MicroShift integrates
with it see following enhancements:
[Integrating MicroShift with Greenboot](https://github.com/openshift/enhancements/blob/master/enhancements/microshift/microshift-greenboot.md)
[MicroShift updateability in ostree based systems: integration with greenboot](https://github.com/openshift/enhancements/blob/master/enhancements/microshift/microshift-updateability-ostree.md#integration-with-greenboot)

In short, it works by running health check scripts
(placed in dirs: `/etc/greenboot/check/{required.d,wanted.d}/`) and, depending
on result of *required* scripts, will run either "green" (healthy) or
"red" (unhealthy) scripts. 

Greenboot heavily integrates with grub. In fact, it is the grub that decrements
the `boot_counter` and actually rolls back the system (by selecting the other
boot target). See [Grub](#grub-ostree) section for more information.

MicroShift delivers these scripts in `microshift-greenboot` RPM package.

Health check script for MicroShift is resides in the repository
as `packaging/greenboot/microshift-running-check.sh`.

> Pre-updateability (and continues to do so), `microshift-greenboot` started to

Reviewer 2: ```suggestion
Before the updateability feature was implemented, `microshift-greenboot` started to
```

provide `packaging/greenboot/microshift-pre-rollback.sh` file as a "red" script.
If `boot_counter` equals `0`, the script will run `microshift-cleanup-data --ovn`.

> Note: It's a red script, so it runs after healthchecks decided that
> system is unhealthy, and cleanup only happens when `boot_counter`
> already fell down to `0`.

### Updateability additions to `microshift-greenboot`

Updateability feature added two, new files - one "green" and one "red" script:
- `packaging/greenboot/microshift_set_healthy.sh` -> `/etc/greenboot/green.d/40_microshift_set_healthy.sh`
- `packaging/greenboot/microshift_set_unhealthy.sh` -> `/etc/greenboot/red.d/40_microshift_set_unhealthy.sh`

They work only on ostree-based systems (by checking if `/run/ostree-booted` exists).
Reason for this is two-fold: how greenboot works on non-ostree systems, and
deployment ID not being available, leaving only boot ID and health, which

> are not as useful without the deployment as backup name.

Reviewer 2: I would simplify this sentence.

Also, since automated backup management is only intended for ostree systems,
there is no gain in creating `health.json` (also: information about the health
can be found in the journal with greater context).

Primary responsibility of "set healthy" and "set unhealthy" scripts
is to create or update `health.json`.

`microshift_set_unhealthy.sh` has an additional logic that, before
updating `health.json`, checks if backup matching deployment and boot IDs
from the `health.json` exists, and does not update the file if not.
This ensures that, when upgrading (from a healthy system) MicroShift fails to
create a backup, in effect causing the system to be unhealthy, that the new health
information does not overwrite existing one ("healthy" instructs MicroShift to

> create a backup).

Reviewer 2: It may be too much detail. Let's simplify and merge the 2 sentences?

Without this mechanism in place, healthy data of previous deployment, would not
get backed up, `health.json` get updated with `unhealthy` causing subsequent
starts of MicroShift want to restore a backup (which might not exist because it
wasn't created, or it's outdated because it contains data of

> "second to last boot of previous deployment").

Reviewer 2: Why do I need to know this information? Seems like an implementation detail to me.

Author: I'm not sure, it is implementation detail, but this is a contributor's doc.
I think it might be useful to at least mention it because I can imagine someone looking up the file and seeing it has `healthy` but the system rebooted and rolled back, but the greenboot didn't finish yet, so the `health.json` has old info. I can even imagine a bug: "health.json contains healthy, even though system was unhealthy"

Reviewer: For the sake of getting all of the "how and why" in one place, let's keep the detail here, for now. If we decide expanding the comments in the script is better we can move some content, but for the short term it's more important to have it written down at all than to get it into exactly the right place.

### Greenboot on non-ostree systems

Although greenboot can be installed and used on non-ostree system, its
usefulness is greatly diminished. It does not perform automated reboots that
try to help system get healthy, nor does it cause a rollback to previously
working system or MicroShift version.

## (rpm-)ostree [ostree-only]

TODO

### `rpm-ostree` vs `ostree`

TODO

### Commits and deployments

TODO

<!--
ostree     -> commits
rpm-ostree -> deployments

Deployment is a commit with a boot entry.
-->

### Useful `ostree` commands

TODO

### Useful `rpm-ostree` commands

TODO

## Grub [ostree]

> While greenboot sets and clears env vars such as `boot_counter` and `boot_success`,

Reviewer 2: ```suggestion
While greenboot sets and clears Grub env vars such as `boot_counter` and `boot_success`,
```

If file existed and version checks were successful, the file is updated 
with version of the executable.

### Kubernetes Storage Migration

Storage migration is process of updating objects to newer versions,
for example from `v1beta1` to `v1beta2`.

Initially, the plan was to perform the migration with only etcd and 
kube-apiserver running.
This turned out to not be a viable options because migrating
CustomResourceDefinitions might require webhooks which run as a Pods, 
therefore requiring kubelet, CNI, DNS, scheduler, and other components.

Right now, storage migration is comprised of two additional controllers
embedded into the MicroShift: Kube Storage Version Trigger and

> Kube Storage Version Migrator. They make use of a `StorageVersionMigration` CR.

Reviewer 3: They make use of the `StorageState` and `StorageVersionMigration` CRs.

> ### `StorageVersionMigration` Custom Resource

Reviewer 3: `StorageState` and `StorageVersionMigration` Custom Resource

It is a Custom Resources created by the Trigger and
picked up to handle by the Migrator.
It can be also used to monitor progress of the migration and get
information about any failures.

#### Kube Storage Version Trigger

Trigger performs a discovery which is a process of gathering information about
existing Kubernetes Resources and their versions.
If the Trigger controller discovers that the Resource should be migrated,
a `StorageVersionMigration` CR is created which is then picked up by 
the Migrator.

> Trigger controller performs the discovery and creates `StorageVersionMigration`

Reviewer 3: A bit more context here with this.

> Trigger controller performs the discovery and creates a `StorageState` and a `StorageVersionMigration` respectively. The `StorageState` contains the current discovered state of the API resource and the previous, if there is a version difference for that specific API resource, then a `StorageVersionMigration` is created by the trigger. This check happens on start and every 10 minutes (discovery period; hardcoded in upstream code).

Author: Thanks! I haven't noticed that one when skimming through the code
